BIP445: FROST Signing Protocol for BIP340 Signatures (#2070)
bitcoin/bips: bip-0445.md

The accompanying source code is licensed under the [MIT license](https://opensource.org/license/mit).

## Motivation

The FROST signature scheme enables threshold Schnorr signatures. In a *t-of-n* threshold configuration, any *t*[^t-edge-cases] participants can cooperatively produce a Schnorr signature that is indistinguishable from a signature produced by a single signer. FROST signatures are unforgeable as long as fewer than *t* participants are corrupted. The signing protocol remains functional provided that at least *t* honest participants retain access to their secret key shares.

[^t-edge-cases]: While *t = n* and *t = 1* are in principle supported, simpler alternatives are available in these cases. In the case *t = n*, using a dedicated *n-of-n* multi-signature scheme such as MuSig2 (see [BIP327][bip327]) instead of FROST avoids the need for an interactive DKG. The case *t = 1* can be realized by letting one signer generate an ordinary [BIP340][bip340] key pair and transmitting the key pair to every other signer, who can check its consistency and then simply use the ordinary [BIP340][bip340] signing algorithm. Signers still need to ensure that they agree on a key pair.

> The IRTF has published [RFC 9591][rfc9591], which specifies the FROST signing protocol for several elliptic curve and hash function combinations, including secp256k1 with SHA-256, the cryptographic primitives used in Bitcoin. However, the signatures produced by RFC 9591 are incompatible with BIP340 Schnorr signatures due to the X-only public keys introduced in BIP340. Additionally, RFC 9591 does not specify key tweaking mechanisms, which are essential for Bitcoin applications such as [BIP32][bip32] key derivation and [BIP341][bip341] Taproot. This document addresses these limitations by specifying a BIP340-compatible variant of FROST signing protocol that supports key tweaking.

Author: The BIP has the following paragraph in the motivation section:

> The FROST3 signing protocol shares substantial similarities with the MuSig2 signing protocol specified in [BIP327][bip327]. Accordingly, this specification adopts several design principles from BIP327, including support for key tweaking, partial signature verification, and identifiable abort mechanisms.

It specifies the key tweaking is adopted from BIP327. I'll now add a note in the algorithms section, about the potential to reuse code from exisiting MuSig2 implementation. This should suffice, right?

> We call it the ["Tweak Context"](https://github.com/bitcoin/bips/blob/ec46a20323840b1a6aba83bc2d18b34dd0811245/bip-frost-signing.md#tweak-context) in this BIP. In BIP327, they call it the ["KeyAgg Context"](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#keyagg-context). At minimum, it seems preferable to use the same nomenclature.

Yes, `TweakContext` is basically the renamed version of `KeyAgg Context`. It would be incorrect to use the same name here because in BIP327, this struct was the output of the `KeyAgg` function (a key generation protocol). However, key generation for this BIP is out of scope, and `TweakContext` is only needed when the user wants to tweak their threshold public key. It's not needed otherwise. Hence, I prefer the `TweakContext` name.

> I would guess that this data structure is gonna become more popular as taproot native protocols will have to handle x-only keys - perhaps its worth having a BIP of its own 🤷‍♂️ .

I agree this structure will become popular. It's already used in [BIP 89](https://github.com/bitcoin/bips/pull/2004/changes#diff-532a50b6ed7a3037322f2093ae38bfe7390c8bba5273e811d54b2682c051bea1). As for standardizing it in a BIP, I think it could be done. Whenever this struct appears, it's usually accompanied by `Tweak Context`, `TweakCtxInit`, and `ApplyTweak`. However, I'm not sure every BIP would follow this exact structure, they may modify it depending on their needs. For instance, this BIP renamed `KeyAggContext`. There's also the possibility of ["agnostic tweaking" in FROST](https://github.com/jesseposner/FROST-BIP340/issues/14), which if adopted would completely eliminate this data structure and would follow a different tweaking approach.

> The IRTF has published [RFC 9591][rfc9591], which specifies the FROST signing protocol for several elliptic curve and hash function combinations, including secp256k1 with SHA-256, the cryptographic primitives used in Bitcoin. However, the signatures produced by RFC 9591 are incompatible with BIP340 Schnorr signatures due to the X-only public keys introduced in BIP340. Additionally, RFC 9591 does not specify key tweaking mechanisms, which are essential for Bitcoin applications such as [BIP32][bip32] key derivation and [BIP341][bip341] Taproot. This document addresses these limitations by specifying a BIP340-compatible variant of FROST signing protocol that supports key tweaking.

Reviewer: The "agnostic tweaking" approach is essentially what is described in [Re-Randomized FROST](https://eprint.iacr.org/2024/436.pdf). You can see in the game in Figure 4 that the attacker provides tweak alpha to the signing oracle, which then tweaks the secret share and public key ("randomize" with alpha) and runs the normal FROST signing algorithm with the tweaked keys.

Btw, this approach does not work with MuSig2. In FROST where `ell_i` and `x_i` denote the lagrange coefficient and share of signer `i` and `t` denote the tweak, tweaking-agnostic signing relies on
```
sum(ell_i(x_i + t)) = sum(ell_i x_i) + t
```
which follows from `sum(ell_i) = 1`. In MuSig2 we'd multiply the tweaked secret key with key aggregation coefficient `a_i`:
```
sum(a_i(x_i + t))
```
which is generally not identical to `sum(a_i x_i) + t`.

> The IRTF has published [RFC 9591][rfc9591], which specifies the FROST signing protocol for several elliptic curve and hash function combinations, including secp256k1 with SHA-256, the cryptographic primitives used in Bitcoin. However, the signatures produced by RFC 9591 are incompatible with BIP340 Schnorr signatures due to the X-only public keys introduced in BIP340. Additionally, RFC 9591 does not specify key tweaking mechanisms, which are essential for Bitcoin applications such as [BIP32][bip32] key derivation and [BIP341][bip341] Taproot. This document addresses these limitations by specifying a BIP340-compatible variant of FROST signing protocol that supports key tweaking.

Author: Agnostic tweaking issue is tracked here: https://github.com/siv2r/bip-frost-signing/issues/41.

> I'll now add a note in the algorithms section, about the potential to reuse code from exisiting MuSig2 implementation. This should suffice, right?

Added a note in `TweakContext` data structure introduction about potential BIP327 code reuse.

| *lift_x(x)*[^liftx-soln] | *GE.lift_x(x)* | Decodes a 32-byte x-only serialization *x* into a non-infinity point P. The resulting point always has an even y-coordinate. |
| *cpoint(b)* | *GE.from_bytes_compressed(b)* | Decodes a 33-byte compressed serialization *b* into a non-infinity point |
| *cpoint_ext(b)* | *GE.from_bytes_compressed<br>_with_infinity(b)* | Decodes a 33-byte compressed serialization *b* into a point. If *b* is a 33-byte array of zeros, it returns the point at infinity |
| *scalar_to_bytes(s)* | *s.to_bytes()* | Returns the 32-byte serialization of a scalar *s* |
| *scalar_from_bytes_checked(b)* | *Scalar.from_bytes_checked(b)* | Deserializes a 32-byte array *b* to a scalar, fails if the value is ≥ *ord* |
| *scalar_from_bytes<br>_nonzero_checked(b)* | *Scalar.from_bytes<br>_nonzero_checked(b)* | Deserializes a 32-byte array *b* to a scalar, fails if the value is zero or ≥ *ord* |
| *scalar_from_bytes_wrapping(b)* | *Scalar.from_bytes_wrapping(b)* | Deserializes a 32-byte array *b* to a scalar, reducing the value modulo *ord* |

> | *hash<sub>tag</sub>(x)* | *tagged_hash(x)* | Computes a 32-byte domain-separated hash of the byte array *x*. The output is *SHA256(SHA256(tag) \|\| SHA256(tag) \|\| x)*, where *tag* is UTF-8 encoded string unique to the context |

Author: Fixed in c408401227803ebd9aa114b6b4471d4a241aa280

### Key Material and Setup

#### Signers Context

The Signers Context is a data structure consisting of the following elements:

- The total number *n* of participants involved in key generation: an integer with *2 ≤ n < 2<sup>32</sup>*
- The threshold number *t* of participants required to issue a signature: an integer with *1 ≤ t ≤ n*
- The number *u* of signing participants: an integer with *t ≤ u ≤ n*
- The list of participant identifiers *id<sub>1..u</sub>*: *u* distinct integers, each with *0 ≤ id<sub>i</sub> ≤ n - 1*
- The list of participant public shares *pubshare<sub>1..u</sub>*: *u* 33-byte arrays, each a compressed serialized point
- The threshold public key *thresh_pk*: a 33-byte array, compressed serialized point

We write "Let *(n, t, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*" to assign names to the elements of Signers Context.

Algorithm *ValidateSignersCtx(signers_ctx)*:

- Inputs:
  - The *signers_ctx*: a [Signers Context](#signers-context) data structure
- *(n, t, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*
- Fail if not *1 ≤ t ≤ n*
- Fail if not *t ≤ u ≤ n*
- For *i = 1 .. u*:

>   - Fail if not *0 ≤ id<sub>i</sub> ≤ n - 1*

Reviewer 2: I posted a response [here](https://github.com/siv2r/bip-frost-signing/issues/9#issuecomment-3947093221), you can resolve this comment if you'd like as I don't think this is worth discussing on this PR.

- If *PartialSigVerifyInternal(psig, my_id, pubnonce, pubshare, session_ctx)* (see below) returns failure, fail[^why-verify-partialsig]
- Return partial signature *psig*

[^why-verify-partialsig]: Verifying the signature before leaving the signer prevents random or adversarially provoked computation errors. This prevents publishing invalid signatures which may leak information about the secret share. It is recommended but can be omitted if the computation cost is prohibitive.

### Partial Signature Verification

Algorithm *PartialSigVerify(psig, pubnonce<sub>1..u</sub>, signers_ctx, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m, i)*:

- Inputs:
  - The partial signature *psig*: a 32-byte array, serialized scalar
  - The list of public nonces *pubnonce<sub>1..u</sub>*: *u* 66-byte arrays, each an output of *NonceGen*
  - The *signers_ctx*: a [Signers Context](#signers-context) data structure
  - The number *v* of tweaks with *0 ≤ v < 2^32*
  - The list of tweaks *tweak<sub>1..v</sub>*: *v* 32-byte arrays, each a serialized scalar
  - The list of tweak modes *is_xonly_t<sub>1..v</sub>* : *v* booleans
  - The message *m*: a byte array[^max-msg-len]

>   - The index *i* of the signer in the list of public nonces where *0 < i ≤ u*

Reviewer 2: Why pass the index here and not the `signer_id` itself? It seems like the index is only used to find the `signer_id`

Also since we start coutning at `0` here, I believe this is off by one (i.e `u+1` elements)?

> 0 < i ≤ u

Author: Fixed in c408401227803ebd9aa114b6b4471d4a241aa280

- Let *(_, _, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, _) = signers_ctx*
- Let *aggnonce = NonceAgg(pubnonce<sub>1..u</sub>, id<sub>1..u</sub>)*; fail if that fails
- Let *session_ctx = (signers_ctx, aggnonce, v, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m)*

> - Run *PartialSigVerifyInternal(psig, id<sub>i</sub>, pubnonce<sub>i</sub>, pubshare<sub>i</sub>, session_ctx)*

Author: Okay. I'll resolve this then.

Implementations may represent this as simply as two separate lists passed to signing APIs.
The threshold public key *thresh_pk* can be stored for efficiency or recomputed when needed using *DeriveThreshPubkey*.
Similarly, the values *n* and *t* are used only for validation, and can be omitted if validation is not performed.

This signing protocol is compatible with any key generation protocol that produces valid FROST keys.
Valid keys satisfy: (1) each *secret share* is a Shamir share of the *threshold secret key*, and (2) each *public share* equals the scalar multiplication *secshare \* G*.
Implementations may **optionally** validate key compatibility for a signing session using the *ValidateSignersCtx* function.
For comprehensive validation of the entire key material, *ValidateSignersCtx* can be run on all possible *u* signing participants.

> [!IMPORTANT]
> Passing *ValidateSignersCtx* ensures functional compatibility with the signing protocol but does not guarantee the security of the key generation protocol itself.

> The output of the FROST signing protocol is a BIP340 Schnorr signature that verifies under the *threshold public key* as if it were produced by a single signer using the *threshold secret key*.

Reviewer 3: slightly more precise (nit? see also https://github.com/bitcoin/bips/pull/2070/changes#diff-d4a124cb94fa5db6244abf10708560c5cdb2c4ba79636316f453f407b0632e5fR67 above).
```suggestion
The output of the FROST signing protocol is a BIP340 Schnorr signature that verifies under the *X-only threshold public key* as if it were produced by a single signer using the *threshold secret key*.
```

Author: Fixed in https://github.com/siv2r/bip-frost-signing/commit/c08325683c22fe087d636947408b6cfd49a6f015

- The total number *n* of participants involved in key generation: an integer with *2 ≤ n < 2<sup>32</sup>*
- The threshold number *t* of participants required to issue a signature: an integer with *1 ≤ t ≤ n*
- The number *u* of signing participants: an integer with *t ≤ u ≤ n*
- The list of participant identifiers *id<sub>1..u</sub>*: *u* distinct integers, each with *0 ≤ id<sub>i</sub> ≤ n - 1*
- The list of participant public shares *pubshare<sub>1..u</sub>*: *u* 33-byte arrays, each a compressed serialized point
- The threshold public key *thresh_pk*: a 33-byte array, compressed serialized point

We write "Let *(n, t, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*" to assign names to the elements of Signers Context.

Algorithm *ValidateSignersCtx(signers_ctx)*:

- Inputs:
  - The *signers_ctx*: a [Signers Context](#signers-context) data structure
- *(n, t, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*
- Fail if not *1 ≤ t ≤ n*
- Fail if not *t ≤ u ≤ n*
- For *i = 1 .. u*:
  - Fail if not *0 ≤ id<sub>i</sub> ≤ n - 1*
  - Fail if *cpoint(pubshare<sub>i</sub>)* fails

> - Fail if *has_duplicates(id<sub>1..u</sub>)*

Reviewer 3: would a duplicate check on the pubshares also make sense here?

Author: Good question. I think the id check is needed but a pubshare one isn't.

The id check matters because computing the interpolating value for a duplicated id would divide by zero (the `id_i - my_id` term). The current code won't actually hit that since it skips when `my_id == id_i`, but it would still derive a wrong key, so we reject duplicates regardless. Pubshares only ever get summed as `lambda_i * P_i`, never used in a denominator, so a repeated pubshare can't break anything.

There are even legitimate cases with repeated pubshares, e.g. for `t=1` every share equals the secret so all pubshares are identical. A duplicate pubshare check would wrongly reject those. So I think it's best to not have the duplicate pubshare check.

- Fail if *DeriveThreshPubkey(id<sub>1..u</sub>, pubshare<sub>1..u</sub>) ≠ thresh_pk*
- No return

Internal Algorithm *DeriveThreshPubkey(id<sub>1..u</sub>,  pubshare<sub>1..u</sub>)*[^derive-thresh-no-validate-inputs]

- *Q = inf_point*
- For *i = 1..u*:

>   - *P* = cpoint(pubshare<sub>i</sub>); fail if that fails

Reviewer 3: the reference implementation throws a `InvalidContributionError` if the compressed point deserialization fails; should a "blame signer at index ..." note be added here accordingly?

Author: Thanks! Added the blame statement in the dev repo.

- If *is_xonly_t* and not *has_even_y(Q)*:
  - Let *g = Scalar(-1)*
- Else:
  - Let *g = Scalar(1)*
- Let *t = scalar_from_bytes_checked(tweak)*; fail if that fails
- Let *Q' = g &middot; Q + t &middot; G*
  - Fail if *is_infinity(Q')*
- Let *gacc' = g &middot; gacc &ensp;(mod ord)*
- Let *tacc' = t + g &middot; tacc &ensp;(mod ord)*
- Return *tweak_ctx' = (Q', gacc', tacc')*

### Nonce Generation

Algorithm *NonceGen(secshare, pubshare, thresh_pk, m, extra_in)*:

- Inputs:
  - The participant secret signing share *secshare*: a 32-byte array, serialized scalar (optional argument)
  - The participant public share *pubshare*: a 33-byte array, compressed serialized point (optional argument)
  - The x-only threshold public key *thresh_pk*: a 32-byte array, X-only serialized point (optional argument). When tweaks have been applied, this is ideally the tweaked threshold public key from *GetXonlyPubkey(tweak_ctx)*.
  - The message *m*: a byte array (optional argument)[^max-msg-len]
  - The auxiliary input *extra_in*: a byte array with *0 ≤ len(extra_in) ≤ 2<sup>32</sup>-1* (optional argument)
- Let *rand' = random_bytes(32)*
- If the optional argument *secshare* is present:

>   - Let *rand = xor_bytes(secshare, hash<sub>FROST/aux</sub>(rand'))*[^sk-xor-rand]

Reviewer 3: Many BIPs that use tagged hashes include the BIP number in the tag (e.g. [BIP324](https://github.com/bitcoin/bips/blob/6deafd07ff5d38c7d69b27530ba55a4ee038cde7/bip-0324/reference.py#L595), [BIP340](https://github.com/bitcoin/bips/blob/6deafd07ff5d38c7d69b27530ba55a4ee038cde7/bip-0340/reference.py#L113), [BIP352](https://github.com/bitcoin/bips/blob/6deafd07ff5d38c7d69b27530ba55a4ee038cde7/bip-0352/reference.py#L97), [BIP374](https://github.com/bitcoin/bips/blob/6deafd07ff5d38c7d69b27530ba55a4ee038cde7/bip-0374/reference.py#L16)). Do we want to do this as well for this one? (as a counter-example though, BIP-327 didn't do it and just uses "MuSig/...", and I guess recreating the test vectors would be at least annoying, so no strong opinion)

Author: Great catch! I was planning to do this after the BIP got a number, but forgot. Thank you!

Author: Fixed in https://github.com/siv2r/bip-frost-signing/commit/c08325683c22fe087d636947408b6cfd49a6f015

As a side effect, the number *n* of signers is not limited by any consensus rules when using FROST.

Moreover, FROST offers a **higher level of privacy** than `OP_CHECKSIGADD`: FROST Taproot outputs are indistinguishable for a blockchain observer from regular, single-signer Taproot outputs even though they are actually controlled by multiple signers. By tweaking the threshold public key, the shared Taproot output can have script spending paths that are hidden unless used.

## Overview

Implementers must make sure to understand this section thoroughly to avoid subtle mistakes that may lead to catastrophic failure.

### Optionality of Features

The goal of this proposal is to support a wide range of possible application scenarios.
Given a specific application scenario, some features may be unnecessary or not desirable, and implementers can choose not to support them.
Such optional features include:

- Applying plain tweaks after x-only tweaks.

> - Not Applying tweaks at all.

Reviewer 4: I noticed that one is negated. Why would "not applying tweaks" be a feature? Wouldn’t either be considered a feature? 

```suggestion
- Applying plain tweaks after x-only tweaks.
- Applying tweaks at all.
```

Author: Good catch, agreed. BIP327 follows this too.

#### Signing Inputs and Outputs

Each signing session requires two inputs: a participant's long-term *secret share* (individual to each participant, not shared with the coordinator) and a [Signers Context](#signers-context)[^signers-ctx-struct] data structure (common to all participants and the coordinator).

[^signers-ctx-struct]: The Signers Context represents the public data of signing participants: their identifiers (*id<sub>1..u</sub>*) and public shares (*pubshare<sub>1..u</sub>*).
Implementations may represent this as simply as two separate lists passed to signing APIs.
The threshold public key *thresh_pk* can be stored for efficiency or recomputed when needed using *DeriveThreshPubkey*.
Similarly, the values *n* and *t* are used only for validation, and can be omitted if validation is not performed.

This signing protocol is compatible with any key generation protocol that produces valid FROST keys.
Valid keys satisfy: (1) each *secret share* is a Shamir share of the *threshold secret key*, and (2) each *public share* equals the scalar multiplication *secshare \* G*.

> Implementations may **optionally** validate key compatibility for a signing session using the *ValidateSignersCtx* function.

Reviewer 5: It was previously discussed in the comments and agreed to be mandatory:

> Duplicate ids in SignersContext means the protocol setup is broken and should be rejected upfront. Now that you mention it, the BIP says validate_signers_context is an "optional" API that implementations can run to check the compatibility of keygen and signing. We should mandate it since we obviously want to avoid duplicate IDs during signing.

I think this was fixed only in the code since the spec still has "optionally".

Author: Fixed in 1f2f91468d2c2041ad72cd51d85a8c708e70c826

#### Signers Context

The Signers Context is a data structure consisting of the following elements:

- The total number *n* of participants involved in key generation: an integer with *2 ≤ n < 2<sup>32</sup>*
- The threshold number *t* of participants required to issue a signature: an integer with *1 ≤ t ≤ n*
- The number *u* of signing participants: an integer with *t ≤ u ≤ n*
- The list of participant identifiers *id<sub>1..u</sub>*: *u* distinct integers, each with *0 ≤ id<sub>i</sub> ≤ n - 1*
- The list of participant public shares *pubshare<sub>1..u</sub>*: *u* 33-byte arrays, each a compressed serialized point
- The threshold public key *thresh_pk*: a 33-byte array, compressed serialized point

We write "Let *(n, t, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*" to assign names to the elements of Signers Context.

Algorithm *ValidateSignersCtx(signers_ctx)*:

- Inputs:
  - The *signers_ctx*: a [Signers Context](#signers-context) data structure
- *(n, t, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*
- Fail if not *1 ≤ t ≤ n*
- Fail if not *t ≤ u ≤ n*
- For *i = 1 .. u*:
  - Fail if not *0 ≤ id<sub>i</sub> ≤ n - 1*

>   - Let *P<sub>i</sub> = cpoint(pubshare<sub>i</sub>)*; fail if that fails and blame signer at index *i* for invalid *pubshare*

Reviewer 5: The code raises `ValueError` here, not `InvalidContributionError`, which is the right call (pre-protocol setup data, not a runtime contribution). The code comment explains this well. The spec text should be updated to just "fail if that fails" to match.

Author: Fixed in 1f2f91468d2c2041ad72cd51d85a8c708e70c826

  Requires: 340
```

## Abstract

This document proposes a standard for the Flexible Round-Optimized Schnorr Threshold (FROST) signing protocol. The standard is compatible with [BIP340][bip340] public keys and signatures. It supports *tweaking*, which allows deriving [BIP32][bip32] child keys from the threshold public key and creating [BIP341][bip341] Taproot outputs with key and script paths.

## Copyright

This document is made available under [CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/).
The accompanying source code is licensed under the [MIT license](https://opensource.org/license/mit).

## Motivation

The FROST signature scheme enables threshold Schnorr signatures. In a *t-of-n* threshold configuration, any *t*[^t-edge-cases] participants can cooperatively produce a Schnorr signature that is indistinguishable from a signature produced by a single signer. FROST signatures are unforgeable as long as fewer than *t* participants are corrupted. The signing protocol remains functional provided that at least *t* honest participants retain access to their secret key shares.

> [^t-edge-cases]: While *t = n* and *t = 1* are in principle supported, simpler alternatives are available in these cases. In the case *t = n*, using a dedicated *n-of-n* multi-signature scheme such as MuSig2 (see [BIP327][bip327]) instead of FROST avoids the need for an interactive DKG. The case *t = 1* can be realized by letting one signer generate an ordinary [BIP340][bip340] key pair and transmitting the key pair to every other signer, who can check its consistency and then simply use the ordinary [BIP340][bip340] signing algorithm. Signers still need to ensure that they agree on a key pair.

Reviewer 6: nit:
```suggestion
[^t-edge-cases]: While *t = n* and *t = 1* are in principle supported, simpler alternatives are available in these cases. In the case *t = n*, using a dedicated *n-of-n* multi-signature scheme such as MuSig2 (see [BIP327][bip327]) instead of FROST avoids the need for an interactive DKG (if using a trusted dealer for key generating is undesirable). The case *t = 1* can be realized by letting one signer generate an ordinary [BIP340][bip340] key pair and transmitting the key pair to every other signer, who can check its consistency and then simply use the ordinary [BIP340][bip340] signing algorithm. Signers still need to ensure that they agree on a key pair.
```
This was taken verbatim from the ChillDKG draft which considers only the DKG case.

Author: Taken as suggested

Following the initial publication of the FROST protocol[[KG20][frost1]], several optimized variants have been proposed to improve computational efficiency and bandwidth optimization: FROST2[[CKM21][frost2]], FROST2-BTZ[[BTZ21][stronger-security-frost]], and FROST3[[RRJSS][roast], [CGRS23][olaf]]. Among these variants, FROST3 is the most efficient variant to date.

This document specifies the FROST3 variant[^frost3-security]. The FROST3 signing protocol shares substantial similarities with the MuSig2 signing protocol specified in [BIP327][bip327]. Accordingly, this specification adopts several design principles from BIP327, including support for key tweaking, partial signature verification, and identifiable abort mechanisms. We note that significant portions of this document have been directly adapted from BIP327 due to the similarities in the signing protocols. Key generation for FROST signing is out of scope for this document.

> [^frost3-security]: The FROST3 signing scheme has been proven existentially unforgeable for both trusted dealer and distributed key generation setups. When using a trusted dealer for key generation, security reduces to the standard One-More Discrete Logarithm (OMDL) assumption. When instantiated with a distributed key generation protocol such as SimplPedPoP, security reduces to the Algebraic One-More Discrete Logarithm (AOMDL) assumption.

Reviewer 6: nit:
```suggestion
[^frost3-security]: FROST3 has been proven existentially unforgeable under the Algebraic One-More Discrete Logarithm (AOMDL) assumption, for both trusted dealer and distributed key generation using the SimplPedPop protocol[[CGRS23][olaf]].
```

It's a bit misleading to distinguish OMDL and AOMDL here.  The Olaf paper shows unforgeability with SimplPedPop under AOMDL (which is a weaker assumption than OMDL and thus prefereable). Unforgeability with SimplPedPop implies unforgeability with a trusted dealer. As a result, unforgeability in both settings reduces to AODML.

Author: Taken as suggested.

This protocol distinguishes between two public key formats: *plain public keys* are 33-byte compressed public keys traditionally used in Bitcoin, while *X-only public keys* are 32-byte keys defined in [BIP340][bip340].
Key generation protocols produce *public shares* and *threshold public keys* in the plain format. During signing, we conditionally negate *secret shares* to ensure the resulting threshold-signature verifies under the corresponding *X-only threshold public key*.

> [!WARNING]
> Key generation protocols must commit the *threshold public key* to an unspendable script path as recommended in [BIP341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-23). This prevents a malicious party from embedding a hidden script path during key generation that would allow them to bypass the *t-of-n* threshold policy.

#### Protocol Parties and Network Setup

> There are *u* (where *t <= u <= n < 2^32*)[^n-bound] participants and one coordinator initiating the FROST signing protocol.

Reviewer 6: This could be interpreted unnecessarily restrictive. In general, there can be *w* participants in the nonce round and *u* in the signing round where *t <= u <= w <= n*. (The most natural case is to request nonces from all n participants and then after you see t of them, pick these for signing; this is exactly what ROAST relies on.) And the algorithms below support *w > u* but someone reading this sentence may think it's now allowed by the BIP. 
 
(Not sure if *w* is good notation. *v* is gone already unfortunately)

Somewhat orthogonal nit: It may be worth adding t <= 1 to the equation in the text here if you mention also the upper bound on n.

Author: Fixed the inequality to 1 <= t <= u <= n < 2^32. For the restrictive reading, I went with a NOTE callout right after this paragraph instead of introducing new notation.

[^signers-ctx-struct]: The Signers Context represents the public data of signing participants: their identifiers (*id<sub>1..u</sub>*) and public shares (*pubshare<sub>1..u</sub>*).
Implementations may represent this as simply as two separate lists passed to signing APIs.
The threshold public key *thresh_pk* can be stored for efficiency or recomputed when needed using *DeriveThreshPubkey*.
Similarly, the values *n* and *t* are used only for validation, and can be omitted if validation is not performed.

This signing protocol is compatible with any key generation protocol that produces valid FROST keys.
Valid keys satisfy: (1) each *secret share* is a Shamir share of the *threshold secret key*, and (2) each *public share* equals the scalar multiplication *secshare \* G*.
Before signing, the signers context must pass *ValidateSignersCtx*, which rejects duplicate identifiers and confirms the key material reproduces the threshold public key. The signing algorithms (*Sign*, *PartialSigVerify*, and *PartialSigAgg*) include this check for clarity, so an implementation can instead validate a context once and skip the repeated checks in later calls.

> For comprehensive validation of the entire key material, *ValidateSignersCtx* can be run on all possible *u* signing participants.

Reviewer 6: nit: Do you mean "on the full set of n participants (u = n)" here? Could be clarified.

edit: On a further look, this is much more sublte. Moved the discussion to the dev repo at https://github.com/siv2r/bip-frost-signing/issues/43#issuecomment-5067697280.

### General Signing Flow

The coordinator and signing participants must be determined before initiating the signing protocol.
The signing participants information is stored in a [Signers Context](#signers-context) data structure.
The *threshold public key* may optionally be tweaked by initializing a [Tweak Context](#tweak-context) at this stage.

Whenever the signing participants want to sign a message, the basic order of operations to create a threshold-signature is as follows:

**First communication round:**
Signers begin the signing session by running *NonceGen* to compute their *secnonce* and *pubnonce*.[^nonce-serialization-detail]
Each signer sends their *pubnonce* to the coordinator, who aggregates them using *NonceAgg* to produce an aggregate nonce and sends it back to all signers.

[^nonce-serialization-detail]: We treat the *secnonce* and *pubnonce* as grammatically singular even though they include serializations of two scalars and two elliptic curve points, respectively.
This treatment may be confusing for readers familiar with the [FROST paper][olaf].

> However, serialization is a technical detail that is irrelevant for users of FROST interfaces.

Reviewer 6: nits:
 - Not just readers of this paper, readers of any FROST paper or anyone familiar with FROST
 - Serialization is a technical detail indeed, but the technical detail whose irrelevance matters here is that the nonce is actually a pair internally.

```suggestion
This treatment may be confusing for readers familiar with the internal workings of FROST.
However, the internal structure of the secnonce and the pubnonce is a technical detail that is irrelevant for users of FROST interfaces.
```

Author: Taken

In most conceivable scenarios, the assumption that the arguments are different between two executions of *NonceGen* is relatively strong, particularly when facing an active adversary.

In some applications, the coordinator may enable preprocessing of nonce generation to reduce signing latency.
Participants run *NonceGen* to generate a batch of *pubnonce* values before the message or Signers Context[^preprocess-round1] is known, which are stored with the coordinator (e.g., on a centralized server).
During this preprocessing phase, only the available arguments are provided to *NonceGen*.
When a signing session begins, the coordinator selects and aggregates *pubnonces* of the signing participants, enabling them to run *Sign* immediately once the message is determined.
This way, the final signature is created quicker and with fewer round trips.
However, applications that use this method presumably store the nonces for a longer time and must therefore be even more careful not to reuse them.
Moreover, this method is not compatible with the defense-in-depth mechanism described in the previous paragraph.

> Generating the nonces ahead of time in this manner does not affect the unforgeability of the scheme.

Reviewer 6: style nit:
"Generating the nonces ahead of time in this manner does not affect the unforgeability of the scheme." appears a bit contradictory to the previous sentences (which stress that you need to be more careful). 
```suggestion
Generating the nonces ahead of time in this manner does not affect the unforgeability of the scheme if nonce reused is properly excluded,
but this method is not compatible with the defense-in-depth mechanism described in the previous paragraph.
```

Author: Taken

> [^det-nonce]: A signer's partial signature has the form *s = k<sub>1</sub> + b k<sub>2</sub> + c &lambda; d*, where *(k<sub>1</sub>, k<sub>2</sub>)* is the secret nonce, *d* is the secret share, and the coefficients *b* (nonce binding), *c* (challenge), and *&lambda;* (Lagrange interpolation over the signer set) are public. With deterministic nonces, an honest signer reproduces the identical *(k<sub>1</sub>, k<sub>2</sub>)* on every signing attempt for a given message. A malicious co-signer exploits this by replaying the session three times on that message and contributing a different nonce each time, which changes the aggregate nonce and therefore both *b* and *c*, while *&lambda;* stays fixed because the signer set is unchanged. The three resulting partial signatures form three linear equations in the unknowns *(k<sub>1</sub>, k<sub>2</sub>, d)*, which the co-signer solves to recover the victim's secret share *d*. This adapts the derandomization attack from Section 3.2 ("Derandomized Signing") of [[MPSW18][musig]], and a similar replay attack is noted for the stateless deterministic signing case in [^det-signer-set].

Reviewer 7: Since `e` is used to denote the challenge later on, it would be better to use `e` instead of `c` here as well, for consistency.

Author: Fixed in https://github.com/siv2r/bip-frost-signing/commit/367695ed0c0f6df09a22b376e00f945f7e86edc4

In particular, partial signatures are *not* signatures.
An adversary can forge a partial signature, i.e., create a partial signature without knowing the secret share for that particular participant public share.[^partialsig-forgery]
However, if *PartialSigVerify* succeeds for all partial signatures then *PartialSigAgg* will return a valid Schnorr signature.

[^partialsig-forgery]: Assume a malicious participant intends to forge a partial signature for the participant with public share *P*. It participates in the signing session pretending to be two distinct signers: one with the public share *P* and the other with its own public share. The adversary then sets the nonce for the second signer in such a way that allows it to generate a partial signature for *P*. As a side effect, it cannot generate a valid partial signature for its own public share. An explanation of the steps required to create a partial signature forgery can be found in [this document](https://gist.github.com/siv2r/0eab97bae9b7186ef2a4919e49d3b426).

### Tweaking the Threshold Public Key

> The threshold public key can be *tweaked*, which modifies the key as defined in the [Tweaking Definition](#tweaking-definition) subsection.

Reviewer 6: It may be worth adding a (foot)note here that for people familiar with MuSig2:

"As opposed to MuSig2, tweaking a single signer's invidual public share (which could be seen as the FROST equivalent to an individual public key) is not meaningful in FROST and thus unsupported."

This is related to my other comment above about the 96-byte secnonce. If you add a note there, it may want to refer to the note here.

Author: Done. Added a [^no-pubshare-tweaking] footnote on the tweaking section's opening sentence

The resulting Tweak Context can be used to apply another tweak with *ApplyTweak* or obtain the threshold public key with *GetXonlyPubkey* or *GetPlainPubkey*.

The purpose of supporting tweaking is to ensure compatibility with existing uses of tweaking, i.e., that the result of signing is a valid signature for the tweaked public key.
The FROST signing algorithms take arbitrary tweaks as input but accepting arbitrary tweaks may negatively affect the security of the scheme.[^arbitrary-tweaks]
Instead, signers should obtain the tweaks according to other specifications.
This typically involves deriving the tweaks from a hash of the threshold public key and some other information.
Depending on the specific scheme that is used for tweaking, either the plain or the X-only threshold public key is required.
For example, to do [BIP32][bip32] derivation, you call *GetPlainPubkey* to be able to compute the tweak, whereas [BIP341][bip341] TapTweaks require X-only public keys that are obtained with *GetXonlyPubkey*.

> [^arbitrary-tweaks]: It is an open question whether allowing *arbitrary* tweaks from an adversary affects the unforgeability of FROST. The standard tweaking used for [BIP32][bip32] derivation and [BIP341][bip341] Taproot corresponds to re-randomizing the threshold public key with an additive tweak, which has been proven to preserve unforgeability under the same AOMDL assumption (in the random oracle model) that underlies plain FROST[[GK24][rerandomized-frost]].

Reviewer 6: ```suggestion
[^arbitrary-tweaks]: It is an open question whether allowing *arbitrary* tweaks from an adversary affects the unforgeability of FROST. The standard tweaking used for [BIP32][bip32] derivation and [BIP341][bip341] Taproot corresponds to re-randomizing the threshold public key with an additive tweak derived via a hash function, which has been proven to preserve unforgeability under the same AOMDL assumption (in the random oracle model) that underlies plain FROST[[GK24][rerandomized-frost]].
```

Author: Taken as suggested

#### Auxiliary and Byte-string Operations

The following helper functions and notation are used for operations on standard integers and byte arrays, independent of curve arithmetic. Note that like Scalars, these variables are denoted by lowercase letters (e.g., *x*, *n*); the intended type is implied by context.

| Notation | Description |
| --- | --- |
| *\|\|* | Refers to byte array concatenation |
| *len(x)* | Returns the length of the byte array *x* in bytes |
| *x[i:j]* | Returns the sub-array of the byte array *x* starting at index *i* (inclusive) and ending at *j* (exclusive). The result has length *j - i* |
| *empty_bytestring* | A constant representing an empty byte array where length is 0 |
| *bytes(n, x)* | Returns the big-endian *n*-byte encoding of the integer *x* |

> | *count(x, lst)* | Returns the number of times the element *x* occurs in the list *lst* |

Reviewer 6: nit: I don't think that one is used at all.

Author: Good catch, removed.

| *has_duplicates(lst)* | Returns *True* if any element in *lst* appears more than once, *False* otherwise |
| *sorted(lst)* | Returns a new list containing the elements of *lst* arranged in ascending order |
| *(a, b, ...)* | Refers to a tuple containing the listed elements |

> [!NOTE]
> In the following algorithms, all scalar arithmetic is understood to be modulo the group order. For example, *a &middot; b* implicitly means *a &middot; b mod order*

### Key Material and Setup

#### Signers Context

The Signers Context is a data structure consisting of the following elements:

> - The total number *n* of participants involved in key generation: an integer with *2 ≤ n < 2<sup>32</sup>*

Reviewer 6: ```suggestion
- The total number *n* of participants involved in key generation: an integer with *1 ≤ n < 2<sup>32</sup>*
```
For consistency with the implemenation.

Author: Fixed, now reads 1 <= n.

[^lagrange-shift]: The standard Lagrange interpolation coefficient uses the formula *id<sub>i</sub> / (id<sub>i</sub> - my_id)* for each term in the product, where ids are in the range *1..n*. However, since participant identifiers in this protocol are zero-indexed (range *0..n-1*), we shift them by adding 1. This transforms each term to *(id<sub>i</sub>+1) / (id<sub>i</sub> - my_id)*.

### Tweaking the Threshold Public Key

#### Tweak Context

The Tweak Context is a data structure consisting of the following elements:

- The point *Q* representing the potentially tweaked threshold public key: a *GE*
- The value *gacc*: *Scalar(1)* or *Scalar(-1)*
- The accumulated tweak *tacc*: a *Scalar*

We write "Let *(Q, gacc, tacc) = tweak_ctx*" to assign names to the elements of a Tweak Context.

> [!TIP]
> The Tweak Context is identical to the *KeyAgg Context* defined in [BIP327][bip327]. Implementations with existing BIP327 code can reuse it to implement this data structure and its algorithms.

Algorithm *TweakCtxInit(thresh_pk):*

- Input:
  - The threshold public key *thresh_pk*: a 33-byte array, compressed serialized point
- Let *Q = cpoint(thresh_pk)*; fail if that fails

> - Fail if *is_infinity(Q)*

Reviewer 6: nit: I believe this failure is unreachable.

Author: Right, cpoint fails first since infinity has no compressed encoding, so this line was unreachable. Removed.

- Let *tacc = Scalar(0)*
- Return *tweak_ctx = (Q, gacc, tacc)*

Algorithm *GetXonlyPubkey(tweak_ctx)*:

- Inputs:
  - The *tweak_ctx*: a [Tweak Context](#tweak-context) data structure
- Let *(Q, _, _) = tweak_ctx*
- Return *xbytes(Q)*

Algorithm *GetPlainPubkey(tweak_ctx)*:

- Inputs:
  - The *tweak_ctx*: a [Tweak Context](#tweak-context) data structure
- Let *(Q, _, _) = tweak_ctx*
- Return *cbytes(Q)*

#### Applying Tweaks

Algorithm *ApplyTweak(tweak_ctx, tweak, is_xonly_t)*:

- Inputs:
  - The *tweak_ctx*: a [Tweak Context](#tweak-context) data structure
  - The *tweak*: a 32-byte array, serialized scalar
  - The tweak mode *is_xonly_t*: a boolean
- Let *(Q, gacc, tacc) = tweak_ctx*
- If *is_xonly_t* and not *has_even_y(Q)*:
  - Let *g = Scalar(-1)*
- Else:
  - Let *g = Scalar(1)*
- Let *t = scalar_from_bytes_checked(tweak)*; fail if that fails
- Let *Q' = g &middot; Q + t &middot; G*
  - Fail if *is_infinity(Q')*
- Let *gacc' = g &middot; gacc &ensp;(mod ord)*
- Let *tacc' = t + g &middot; tacc &ensp;(mod ord)*
- Return *tweak_ctx' = (Q', gacc', tacc')*

### Nonce Generation

Algorithm *NonceGen(secshare, pubshare, thresh_pk, m, extra_in)*:

- Inputs:

>   - The participant secret signing share *secshare*: a 32-byte array, serialized scalar (optional argument)

Reviewer 6: micro/consistency nit:
These are called "secret share" everywhere in the BIP except for two places where they're called "secret signing share". Intentional?

Author: Not intentional. Unified all four occurrences to "secret share".

  - The participant public share *pubshare*: a 33-byte array, compressed serialized point (optional argument)

>   - The x-only threshold public key *thresh_pk*: a 32-byte array, X-only serialized point (optional argument). When tweaks have been applied, this is ideally the tweaked threshold public key from *GetXonlyPubkey(tweak_ctx)*.

Reviewer 6: This is in the list of inputs to NonceGen.

It was a bit confusing to me that here the algorithms expect an xonly threshold pubkey as input whereas the signers_ctx requires a plain pubkey. (In particular because they use the same identifier `thresh_pk`.)

As far as I understand, the goal is to take account the tweaks for defense-in-depth reason (which is a good idea and an improvement over BIP327!).

I see two possible improvements:
 - Make the wording and notation a bit clearer (see below for a possible wording).
 - Or, instead of taking the xonly key, take all tweaks (and is_xonly_t values) and hash them.

```suggestion
  - The x-only threshold public key *tweaked_xonly_thresh_pk*: a 32-byte array, X-only serialized point (optional argument). If tweaks are supposed to be applied during signing, this should ideally be the resulting xonly public key after applying all tweaks to the threshold public key, i.e., the output of *GetXonlyPubkey(tweak_ctx)*. If no tweaks are supposed to be applied, or if it is not yet determined at nonce generation stage whether or what tweaks will be applied, this may also be the last 32 bytes of the threshold public key. 
```

Author: I thought about your second option, hashing all tweaks and `is_xonly_t` values. I believe it doesn't add any protection over hashing a single tweaked key, and the current x-only input is exactly as strong as both. So the choice is purely an API question. Here's my thinking:

All signing algorithms consume the tweak list only through the triple $(Q_v, gacc_v, tacc_v)$, which satisfies the invariant $Q_v = gacc_v \cdot Q_0 + tacc_v \cdot G$ with $gacc_v \in \{1, -1\}$. Concretely, $b$ and $e$ hash $xbytes(Q_v)$, *Sign* computes $d = g \cdot gacc \cdot d'$, *PartialSigAgg* adds $e \cdot g \cdot tacc$, and $g$ is determined by the parity of $Q_v$.

Let's consider two different tweak sequences applied to the same threshold key and analyze when they collide on the *NonceGen* hash. Assume $rand'$ is identical in both runs, since that's the failure mode the defense-in-depth targets. Note that since $gacc \in \{1, -1\}$, the two sequences either have the same $gacc$ or opposite ones, so there are only two cases per option.

Option 1: *NonceGen* hashes the plain tweaked key (`PlainPk`). A collision means the same $Q_v$.

- Case 1 (same $gacc$): the invariant forces the same $tacc$, so both contexts produce byte-identical partial signatures. Deriving the same nonce for them is harmless.
- Case 2 (opposite $gacc$): the invariant forces $tacc' = tacc + 2 \cdot gacc \cdot x_0$, where $x_0$ is the threshold secret key. Constructing such a pair requires knowing $x_0$, and honest BIP32/BIP341 hash-derived tweaks hit it with probability $\sim 2^{-256}$.

Option 2: *NonceGen* hashes the tweak and tweak mode lists (with count prefixing). Distinct sequences always get distinct nonces. The only pairs this separates beyond Option 1 are the benign ones from Case 1, so it adds no extra protection.

Option 3: *NonceGen* hashes the x-only tweaked key (`XonlyPk`, current spec). A collision means $Q_v' = \pm Q_v$, giving four cases.

- Case 1 (same $gacc$, same $Q_v$): same as Option 1, Case 1.
- Case 2 (same $gacc$, opposite $Q_v$): the invariant forces $tacc' = -tacc - 2 \cdot gacc \cdot x_0$, so it requires knowing $x_0$, same reasoning as Option 1, Case 2.
- Case 3 (opposite $gacc$, same $Q_v$): the invariant forces $tacc' = tacc + 2 \cdot gacc \cdot x_0$, which is exactly Option 1, Case 2.
- Case 4 (opposite $gacc$, opposite $Q_v$): the invariant forces $tacc' = -tacc$. Here $g$ flips together with the parity of $Q_v$, so $g \cdot gacc$ and $g \cdot tacc$ are unchanged, and both contexts produce byte-identical signatures for the same x-only key. Benign, like Case 1.

So all three options give equally good defense-in-depth. Given that, I'd prefer switching the argument from x-only to the plain, potentially tweaked key, i.e. the output of *GetPlainPubkey(tweak_ctx)*. It matches the `thresh_pk` field in the Signers Context, matches what key generation outputs (ChillDKG returns a plain threshold pubkey), and lets the untweaked case pass `thresh_pk` through unchanged, which removes the confusion you pointed out. Plus the change is quite minimal compared to adding two new list params to *NonceGen*.

Reviewer 6: > So all three options give equally good defense-in-depth

That was my assumption, too. (I haven't verified your entire analysis.)

> Given that, I'd prefer switching the argument from x-only to the plain, potentially tweaked key, i.e. the output of _GetPlainPubkey(tweak_ctx)_. It matches the `thresh_pk` field in the Signers Context, matches what key generation outputs (ChillDKG returns a plain threshold pubkey), and lets the untweaked case pass `thresh_pk` through unchanged, which removes the confusion you pointed out. Plus the change is quite minimal compared to adding two new list params to _NonceGen_.

While this removes could remove a bit of the confusion, the caller may not have the plain threshold pubkey (and this may actually add to the confusion again.) I think it's reasonable to assume that a wallet, which tries to spend a UTXO, knows (or has cached) the tweaked xonly pubkey (because this is literally in the UTXO), in addition to the untweaked threshold public key and the tweaks (which are required for signing anyway). But the wallet may not know the missing sign byte of the tweaked pubkey because it's not in the UTXO.

So an API that expects a plain final pubkey could make the caller believe that it needs to recompute the tweaked plain pubkey before nonce generation (which is unnecessarily expensive given that it happens during signing anyway, and also may lead to the decision not to pass this argument at all to nonce generation). 

From that point of view, an API that expects the tweaked xonly pubkey in nonce generation is actually quite natural. (Well, as far as the entire xonly business is natural.) And I tend to say, let's stick to it and simply make the docs clear here.

A forth option is make the argument 33 bytes and tell the caller that it can either pass a plain pubkey or an xonly pubkey prefixed with a zero. But that doesn't sound great either.

Author: Okay. Sounds good. I'll just make the docs more clear.

Author: Fixed in f7de456

- Let *rand' = random_bytes(32)*
- If the optional argument *secshare* is present:
  - Let *rand = xor_bytes(secshare, hash<sub>BIP0445/aux</sub>(rand'))*[^sk-xor-rand]
- Else:
  - Let *rand = rand'*
- If the optional argument *pubshare* is not present:
  - Let *pubshare* = *empty_bytestring*
- If the optional argument *thresh_pk* is not present:
  - Let *thresh_pk* = *empty_bytestring*
- If the optional argument *m* is not present:
  - Let *m_prefixed = bytes(1, 0)*
- Else:
  - Let *m_prefixed = bytes(1, 1) || bytes(8, len(m)) || m*
- If the optional argument *extra_in* is not present:
  - Let *extra_in = empty_bytestring*
- Let *k<sub>i</sub> = scalar_from_bytes_wrapping(hash<sub>BIP0445/nonce</sub>(rand || bytes(1, len(pubshare)) || pubshare || bytes(1, len(thresh_pk)) || thresh_pk || m_prefixed || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1)))* for *i = 1,2*
- Fail if *k<sub>1</sub> = Scalar(0)* or *k<sub>2</sub> = Scalar(0)*[^negligible-zero-scalar]
- Let *R<sub>\*,1</sub> = k<sub>1</sub> &middot; G*, *R<sub>\*,2</sub> = k<sub>2</sub> &middot; G*
- Let *pubnonce = cbytes(R<sub>\*,1</sub>) || cbytes(R<sub>\*,2</sub>)*

> - Let *secnonce = scalar_to_bytes(k<sub>1</sub>) || scalar_to_bytes(k<sub>2</sub>)*[^secnonce-ser]

Reviewer 6: It may be worth adding a footnote on the difference to BIP327 here. Suggestion:

"BIP 327 appends the serialized individual pk to the secnonce (resulting in a 96-byte secnonce) to avoid a vulnerability that may arise when MuSig2 signers tweak their individual key pair before key aggregation. In FROST, the threshold public key is fixed per FROST key generation and tweaking the public share (which could be seen as the FROST equivalent to an individual public key) is not supported. Thus, this vulnerability does not apply to FROST, and appending the public share to the secnonce is not necessary.

Author: Done. Added a [^secnonce-vs-bip327] footnote on the secnonce line explaining why 64 bytes suffice (the threshold key is fixed at keygen and pubshare tweaking is unsupported), cross-referencing the pubshare-tweaking note above.

[^max-msg-len]: In theory, the allowed message size is restricted because SHA256 accepts byte strings only up to size of 2^61-1 bytes (and because of the 8-byte length encoding).

[^negligible-zero-scalar]: These are unreachable errors, included for completeness: such a value equals *Scalar(0)* only with negligible probability. The reference implementation checks the condition with an assertion.

### Nonce Aggregation

Algorithm *NonceAgg(pubnonce<sub>1..u</sub>)*:

- Inputs:
  - The number *u* of signing participants: an integer with *t ≤ u ≤ n*

>   - The list of participant public nonces *pubnonce<sub>1..u</sub>*: *u* 66-byte array, each an output of *NonceGen*

Reviewer 6: nit:
```suggestion
  - The list of participant public nonces *pubnonce<sub>1..u</sub>*: *u* 66-byte arrays each an output of *NonceGen*
```

- For *j = 1 .. 2*:
  - For *i = 1 .. u*:
    - Let *R<sub>i,j</sub> = cpoint(pubnonce<sub>i</sub>[(j-1)\*33:j\*33])*; fail if that fails and blame signer at index *i* for invalid *pubnonce*
  - Let *R<sub>j</sub> = R<sub>1,j</sub> + R<sub>2,j</sub> + ... + R<sub>u,j</sub>*
- Return *aggnonce = cbytes_ext(R<sub>1</sub>) || cbytes_ext(R<sub>2</sub>)*

### Session Context

The Session Context is a data structure consisting of the following elements:

- The *signers_ctx*: a [Signers Context](#signers-context) data structure
- The aggregate public nonce *aggnonce*: a 66-byte array, output of *NonceAgg*
- The number *v* of tweaks with *0 ≤ v < 2^32*
- The list of tweaks *tweak<sub>1..v</sub>*: *v* 32-byte arrays, each a serialized scalar
- The list of tweak modes *is_xonly_t<sub>1..v</sub>* : *v* booleans
- The message *m*: a byte array[^max-msg-len]

> We write "Let *(signers_ctx, aggnonce, v, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m) = session_ctx*" to assign names to the elements of a Session Context.

Reviewer 6: nit: Claude says that the order in the code is different: `(aggnonce, signers_ctx, tweaks, is_xonly, msg)`. Making pseudocode and consistent helps a reader comparing the two.

Author: Fixed

Algorithm *GetSessionValues(session_ctx)*:

- Let *(signers_ctx, aggnonce, v, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m) = session_ctx*
- *ValidateSignersCtx(signers_ctx)*; fail if that fails
- Let *(_, _, u, id<sub>1..u</sub>, pubshare<sub>1..u</sub>, thresh_pk) = signers_ctx*
- Let *tweak_ctx<sub>0</sub> = TweakCtxInit(thresh_pk)*; fail if that fails
- For *i = 1 .. v*:
  - Let *tweak_ctx<sub>i</sub> = ApplyTweak(tweak_ctx<sub>i-1</sub>, tweak<sub>i</sub>, is_xonly_t<sub>i</sub>)*; fail if that fails
- Let *(Q, gacc, tacc) = tweak_ctx<sub>v</sub>*
- Let *ser_ids* = *SerializeIds(id<sub>1..u</sub>)*[^canonical-ids-det-sign]

> - Let *b* = *scalar_from_bytes_wrapping(hash<sub>BIP0445/noncecoef</sub>(ser_ids || aggnonce || xbytes(Q) || m))*

Reviewer: `ser_ids` and `m` are of arbitrary length and not length prefixed when input to the `noncecoef` hash function. Hence, it seems easy to find tuples `(ids, m) != (ids', m')` that result in the same hash `b`.

Reviewer 6: Oh, "nice" find. A simple fix is to let `SerializeIds` prefix its output with a serialization of `u`.

edit: Or, for consistency, make it similar to how the inputs to *hash<sub>BIP0445/deterministic/nonce</sub>* are serialized.

Reviewer: Found by ChatGPT 5.6 when trying to prove security of the scheme specified in the BIP.

Author: Thanks for pointing this out! Fixed in f7de456. The hash input now has a signers count prefix.

The security of the resulting scheme then depends on the requirement that reading the counter must never yield the same counter value in two *NonceGen* invocations with the same *secshare*.

Second, if there is a unique signer who generates their nonce last (i.e., after receiving the aggregate nonce from all other signers), it is possible to modify nonce generation for this single signer to not require high-quality randomness.
Such a nonce generation algorithm *DeterministicSign* is specified below.
It has two optional arguments: *rand*, which can be omitted if randomness is entirely unavailable, and *aggothernonce*, which is omitted by a sole signer (*u = 1*) who has no other signers' nonces to aggregate.
When present, *aggothernonce* should be set to the output of *NonceAgg* run on the *pubnonce* value of **all** other signers (but can be provided by an untrusted party).
Hence, using *DeterministicSign* is only possible for the last signer to generate a nonce, or for a sole signer who is the only participant signing, and it makes the signer stateless, similar to the stateless signer described in the [Nonce Generation](#nonce-generation) section.

> In FROST, the deterministic nonce must also bind to the the signer subset *id<sub>1..u</sub>*; otherwise a malicious coordinator can recover the victim's secret share via replayed sessions with varying signer subsets.[^det-signer-set]

Reviewer 6: nit: double "the"
```suggestion
In FROST, the deterministic nonce must also bind to the signer subset *id<sub>1..u</sub>*; otherwise a malicious coordinator can recover the victim's secret share via replayed sessions with varying signer subsets.[^det-signer-set]
```

Author: Fixed

  - Add Python script to automate generation of test vectors.
  - Represent participant identifiers as 4-byte integers in the range *0..n - 1* (inclusive).
- *0.1.0* (2024-07-31): Publication of draft BIP on the bitcoin-dev mailing list

## Acknowledgments

We thank Jonas Nick, Tim Ruffing, Jesse Posner, Sebastian Falbesoner, Chris Stewart, Illia Melnyk, Mariia Zhvanko, and Fadi Barbàra for their comments and contributions to this document.

<!-- References -->
[bip32]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
[bip340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
[bip341]: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki
[bip342]: https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki
[bip327]: https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki
[musig]: https://eprint.iacr.org/2018/068
[frost1]: https://eprint.iacr.org/2020/852
[frost2]: https://eprint.iacr.org/2021/1375
[stronger-security-frost]: https://eprint.iacr.org/2022/833
[olaf]: https://eprint.iacr.org/2023/899
[roast]: https://eprint.iacr.org/2022/550
[rerandomized-frost]: https://eprint.iacr.org/2024/436

> <!-- [thresh-with-dkg]: https://link.springer.com/chapter/10.1007/3-540-36563-x_26 -->

Reviewer 6: You may want to remove this line if it's unused.

Author: Removed
